File: server/controllers/authController.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const User = require('../models/User');
require('dotenv').config();

const generateToken = (user) => {
    return jwt.sign(
        { id: user._id, email: user.email, name: user.name, role: user.role || 'user' },
        process.env.JWT_SECRET,
        { expiresIn: process.env.JWT_EXPIRE || '7d' }
    );
};

// @route   POST /api/auth/register
exports.register = async (req, res) => {
    try {
        const { name, email, password, role = 'user', organisation } = req.body;

        // Validation
        if (!name || !email || !password) {
            return res.status(400).json({
                success: false,
                message: 'Please provide name, email, and password'
            });
        }

        if (password.length < 6) {
            return res.status(400).json({
                success: false,
                message: 'Password must be at least 6 characters'
            });
        }

        // Only allow 'user' or 'municipal_admin' roles on self-registration
        const allowedRoles = ['user', 'municipal_admin'];
        if (!allowedRoles.includes(role)) {
            return res.status(400).json({
                success: false,
                message: 'Invalid role'
            });
        }

        // Municipal authority registration requires organisation
        if (role === 'municipal_admin' && !organisation) {
            return res.status(400).json({
                success: false,
                message: 'Municipality/Organisation name is required for authority registration'
            });
        }

        // Check if user exists
        const existingUser = await User.findOne({ email: email.toLowerCase() });

        if (existingUser) {
            return res.status(400).json({
                success: false,
                message: 'User with this email already exists'
            });
        }

        // Hash password
        const salt = await bcrypt.genSalt(12);
        const hashedPassword = await bcrypt.hash(password, salt);

        // Municipal admins start as pending until super admin approves
        const approvalStatus = role === 'municipal_admin' ? 'pending' : 'approved';

        // Create user
        const user = await User.create({
            name,
            email: email.toLowerCase(),
            password: hashedPassword,
            role,
            approval_status: approvalStatus,
            organisation: organisation || null
        });

        const userData = {
            id: user._id,
            name: user.name,
            email: user.email,
            role: user.role,
            approval_status: user.approval_status,
            organisation: user.organisation,
            contribution_score: user.contribution_score,
            reports_count: user.reports_count,
            verified_reports: user.verified_reports,
            badges: user.badges,
            created_at: user.created_at
        };

        // For municipal_admin pending approval, don't issue token - they need to wait
        if (role === 'municipal_admin' && approvalStatus === 'pending') {
            return res.status(201).json({
                success: true,
                message: 'Registration submitted! Your account is pending approval by the Super Admin. You will be able to log in once approved.',
                data: {
                    user: { id: user._id, name: user.name, email: user.email, role: user.role, approval_status: user.approval_status },
                    token: null,
                    pending: true
                }
            });
        }

        const token = generateToken(user);

        res.status(201).json({
            success: true,
            message: 'Registration successful',
            data: {
                user: userData,
                token
            }
        });
    } catch (error) {
        console.error('Register error:', error);
        res.status(500).json({
            success: false,
            message: 'Server error during registration'
        });
    }
};

// @route   POST /api/auth/login
exports.login = async (req, res) => {
    try {
        const { email, password } = req.body;

        if (!email || !password) {
            return res.status(400).json({
                success: false,
                message: 'Please provide email and password'
            });
        }

        // Find user
        const user = await User.findOne({ email: email.toLowerCase() });

        if (!user) {
            return res.status(401).json({
                success: false,
                message: 'Invalid credentials'
            });
        }

        // Verify password
        const isMatch = await bcrypt.compare(password, user.password);
        if (!isMatch) {
            return res.status(401).json({
                success: false,
                message: 'Invalid credentials'
            });
        }

        // Block pending municipal admins
        if (user.role === 'municipal_admin' && user.approval_status === 'pending') {
            return res.status(403).json({
                success: false,
                message: 'Your municipal authority account is pending approval by the Super Admin. Please wait for approval before logging in.',
                pending: true
            });
        }

        // Block rejected municipal admins
        if (user.role === 'municipal_admin' && user.approval_status === 'rejected') {
            return res.status(403).json({
                success: false,
                message: 'Your municipal authority account registration was rejected. Please contact the administrator.',
                rejected: true
            });
        }

        const token = generateToken(user);

        res.json({
            success: true,
            message: 'Login successful',
            data: {
                user: {
                    id: user._id,
                    name: user.name,
                    email: user.email,
                    role: user.role || 'user',
                    approval_status: user.approval_status,
                    organisation: user.organisation,
                    contribution_score: user.contribution_score,
                    reports_count: user.reports_count,
                    verified_reports: user.verified_reports,
                    badges: user.badges || [],
                    created_at: user.created_at
                },
                token
            }
        });
    } catch (error) {
        console.error('Login error:', error);
        res.status(500).json({
            success: false,
            message: 'Server error during login'
        });
    }
};

// @route   GET /api/auth/me
exports.getMe = async (req, res) => {
    try {
        const user = await User.findById(req.user.id).select('-password');

        if (!user) {
            return res.status(404).json({
                success: false,
                message: 'User not found'
            });
        }

        res.json({
            success: true,
            data: {
                id: user._id,
                name: user.name,
                email: user.email,
                role: user.role,
                approval_status: user.approval_status,
                organisation: user.organisation,
                avatar_url: user.avatar_url,
                contribution_score: user.contribution_score,
                reports_count: user.reports_count,
                verified_reports: user.verified_reports,
                badges: user.badges || [],
                created_at: user.created_at
            }
        });
    } catch (error) {
        console.error('GetMe error:', error);
        res.status(500).json({
            success: false,
            message: 'Server error'
        });
    }
};

File: server/controllers/auth_controller.py
import os
import jwt
import bcrypt
from datetime import datetime, timedelta
from fastapi import HTTPException
from bson import ObjectId
from config.db import get_db
from utils.helpers import serialize_doc

JWT_SECRET = os.getenv("JWT_SECRET")
JWT_EXPIRE_DAYS = 7 # Default matching '7d'

def generate_token(user_id: str, email: str, name: str, role: str) -> str:
    payload = {
        "id": user_id,
        "email": email,
        "name": name,
        "role": role,
        "exp": datetime.utcnow() + timedelta(days=JWT_EXPIRE_DAYS)
    }
    return jwt.encode(payload, JWT_SECRET, algorithm="HS256")

async def register_user(body: dict):
    name = body.get("name")
    email = body.get("email")
    password = body.get("password")
    role = body.get("role", "user")
    organisation = body.get("organisation")

    if not name or not email or not password:
        raise HTTPException(status_code=400, detail="Please provide name, email, and password")

    if len(password) < 6:
        raise HTTPException(status_code=400, detail="Password must be at least 6 characters")

    allowed_roles = ["user", "municipal_admin"]
    if role not in allowed_roles:
        raise HTTPException(status_code=400, detail="Invalid role")

    if role == "municipal_admin" and not organisation:
        raise HTTPException(status_code=400, detail="Municipality/Organisation name is required for authority registration")

    db = get_db()
    existing_user = await db["users"].find_one({"email": email.lower()})
    if existing_user:
        raise HTTPException(status_code=400, detail="User with this email already exists")

    salt = bcrypt.gensalt(12)
    hashed_password = bcrypt.hashpw(password.encode('utf-8'), salt).decode('utf-8')
    approval_status = "pending" if role == "municipal_admin" else "approved"

    new_user = {
        "name": name,
        "email": email.lower(),
        "password": hashed_password,
        "role": role,
        "approval_status": approval_status,
        "organisation": organisation if role == "municipal_admin" else None,
        "avatar_url": None,
        "contribution_score": 0,
        "reports_count": 0,
        "verified_reports": 0,
        "badges": [],
        "created_at": datetime.utcnow(),
        "updated_at": datetime.utcnow()
    }

    result = await db["users"].insert_one(new_user)
    user_id = str(result.inserted_id)

    if role == "municipal_admin" and approval_status == "pending":
        return {
            "success": True,
            "message": "Registration submitted! Your account is pending approval by the Super Admin. You will be able to log in once approved.",
            "data": {
                "user": {"id": user_id, "name": name, "email": email.lower(), "role": role, "approval_status": approval_status},
                "token": None,
                "pending": True
            }
        }

    token = generate_token(user_id, email.lower(), name, role)
    user_data = {
        "id": user_id,
        "name": name,
        "email": email.lower(),
        "role": role,
        "approval_status": approval_status,
        "organisation": new_user["organisation"],
        "contribution_score": 0,
        "reports_count": 0,
        "verified_reports": 0,
        "badges": [],
        "created_at": new_user["created_at"]
    }

    return {
        "success": True,
        "message": "Registration successful",
        "data": {
            "user": serialize_doc(user_data),
            "token": token
        }
    }

async def login_user(body: dict):
    email = body.get("email")
    password = body.get("password")

    if not email or not password:
        raise HTTPException(status_code=400, detail="Please provide email and password")

    db = get_db()
    user = await db["users"].find_one({"email": email.lower()})
    if not user:
        raise HTTPException(status_code=401, detail="Invalid credentials")

    # verify password
    if not bcrypt.checkpw(password.encode('utf-8'), user["password"].encode('utf-8')):
        raise HTTPException(status_code=401, detail="Invalid credentials")

    if user.get("role") == "municipal_admin":
        if user.get("approval_status") == "pending":
            raise HTTPException(
                status_code=403, 
                detail="Your municipal authority account is pending approval by the Super Admin. Please wait for approval before logging in."
            )
        if user.get("approval_status") == "rejected":
            raise HTTPException(
                status_code=403,
                detail="Your municipal authority account registration was rejected. Please contact the administrator."
            )

    token = generate_token(str(user["_id"]), user["email"], user["name"], user.get("role", "user"))
    
    user_data = {
        "id": str(user["_id"]),
        "name": user["name"],
        "email": user["email"],
        "role": user.get("role", "user"),
        "approval_status": user.get("approval_status", "approved"),
        "organisation": user.get("organisation"),
        "contribution_score": user.get("contribution_score", 0),
        "reports_count": user.get("reports_count", 0),
        "verified_reports": user.get("verified_reports", 0),
        "badges": user.get("badges", []),
        "created_at": user.get("created_at")
    }

    return {
        "success": True,
        "message": "Login successful",
        "data": {
            "user": serialize_doc(user_data),
            "token": token
        }
    }

async def get_me(user_id: str):
    db = get_db()
    user = await db["users"].find_one({"_id": ObjectId(user_id)}, {"password": 0})
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
        
    return {
        "success": True,
        "data": serialize_doc(user)
    }
